webrtc-c/canary/src/CanarySignaling.cpp

```diff
@@ -1,6 +1,522 @@
+{
+    // Prepare the test channel name by prefixing with test channel name
+    // and generating random chars replacing a potentially bad characters with '.'
+    STRCPY(pChannelName, SIGNALING_CANARY_CHANNEL_NAME);
+    UINT32 nameLen = STRLEN(SIGNALING_CANARY_CHANNEL_NAME);
+    const UINT32 randSize = 16;
+
+    PCHAR pCur = &pChannelName[nameLen];
+
+    for (UINT32 i = 0; i < randSize; i++) {
+        *pCur++ = SIGNALING_VALID_NAME_CHARS[RAND() % (ARRAY_SIZE(SIGNALING_VALID_NAME_CHARS) - 1)];
+    }
+
+    *pCur = '\0';
+}
+
+STATUS run(Canary::PConfig pConfig)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL initialized = FALSE;
+    TIMER_QUEUE_HANDLE timerQueueHandle = 0;
+    UINT32 timeoutTimerId;
+    ChannelInfo masterChannelInfo = {0};
+    ChannelInfo viewerChannelInfo = {0};
+    SignalingClientInfo masterClientInfo = {0};
+    SignalingClientInfo viewerClientInfo = {0};
+    SignalingClientCallbacks masterSignalingClientCallbacks = {0};
+    SignalingClientCallbacks viewerSignalingClientCallbacks = {0};
+    PAwsCredentialProvider pCredentialProvider = NULL;
+    SIGNALING_CLIENT_HANDLE masterSignalingClientHandle = INVALID_SIGNALING_CLIENT_HANDLE_VALUE;
+    SIGNALING_CLIENT_HANDLE viewerSignalingClientHandle = INVALID_SIGNALING_CLIENT_HANDLE_VALUE;
+    CanarySessionInfo canarySessionInfo = {0};
+    MUTEX lock = INVALID_MUTEX_VALUE;
+    CVAR terminateCv = INVALID_CVAR_VALUE;
+    CHAR channelName[MAX_CHANNEL_NAME_LEN + 1];
+
+    canarySessionInfo.roundtripLock = INVALID_MUTEX_VALUE;
+    canarySessionInfo.roundtripCv = INVALID_CVAR_VALUE;
+
+    // Create the lock and the cvar for iteration
+    CHK(IS_VALID_MUTEX_VALUE(lock = MUTEX_CREATE(FALSE)), STATUS_NOT_ENOUGH_MEMORY);
+    CHK(IS_VALID_CVAR_VALUE(terminateCv = CVAR_CREATE()), STATUS_NOT_ENOUGH_MEMORY);
+    CHK(IS_VALID_MUTEX_VALUE(canarySessionInfo.roundtripLock = MUTEX_CREATE(FALSE)), STATUS_NOT_ENOUGH_MEMORY);
+    CHK(IS_VALID_CVAR_VALUE(canarySessionInfo.roundtripCv = CVAR_CREATE()), STATUS_NOT_ENOUGH_MEMORY);
+    canarySessionInfo.exitStatus = STATUS_SUCCESS;
+    canarySessionInfo.iterationFailCount = 0;
+
+    CHK_STATUS(Canary::Cloudwatch::init(pConfig));
+    CHK_STATUS(initKvsWebRtc());
+    initialized = TRUE;
+
+    SET_LOGGER_LOG_LEVEL(pConfig->logLevel);
+    pConfig->print();
+
+    // The timer loop for iteration
+    CHK_STATUS(timerQueueCreate(&timerQueueHandle));
+
+    // We will create a static credential provider. We can replace it with others if needed.
+    CHK_STATUS(createStaticCredentialProvider((PCHAR) pConfig->pAccessKey, 0,
+                                              (PCHAR) pConfig->pSecretKey, 0,
+                                              (PCHAR) pConfig->pSessionToken, 0,
+                                              MAX_UINT64,
+                                              &pCredentialProvider));
+
+    // Generate a random channel name
+    generateChannelName(channelName);
```

Comment:
You are right. Let me add logic that has the best of both of the worlds. If the channel name is specified then it will be used and the canary won't attempt to delete it afterwards. If, however, its not specified then it will generate, then delete it